koku/api/query_params.py

```diff
@@ -425,25 +387,61 @@ def _set_access_ocp_all(self, provider, filter_key, access_key, raise_exception=
             elif access_list:
                 self.parameters["filter"][filter_key] = access_list
 
-    def _set_tag_keys(self):
+    def search_url_for_prefixes(self, prefix_list):
+        """Searches url and returns true if any prefix is found in a url."""
+        # the url_data property is percent encoded
+        # https://en.wikipedia.org/wiki/URL_encoding
+        # %5Btag == [tag:
+        # %5Bor%3Atag == [or:tag
+        decoded_url = unquote(self.url_data)
+        for prefix in prefix_list:
+            if f"[{prefix}" in decoded_url:
+                return True
+        return False
+
+    def _set_tag_keys(self, query_params):
         """Set the valid tag keys"""
+        prefix_list = [TAG_PREFIX, OR_TAG_PREFIX, AND_TAG_PREFIX]
         self.tag_keys = set()
-        # we do not need to fetch the tags for tags report type.
-        # we also do not need to fetch the tags if none of the query params contain `tags`
-        if self.report_type == "tags" or "tag" not in self.url_data:
+        if self.report_type == "tags" or not self.search_url_for_prefixes(prefix_list):
+            # we do not need to fetch the tags for tags report type.
+            # we also do not need to fetch the tags if a tag prefix is not in the URL
             return
-
         for tag_model in self.tag_handler:
             with tenant_context(self.tenant):
                 self.tag_keys.update(tag_model.objects.values_list("key", flat=True).distinct())
+        param_tag_keys = set()
+        # Reduce the set of tag keys based on those being queried.
+        for key, value in query_params.items():
+            if not isinstance(value, (dict, list)):
+                value = [value]
+            for inner_key in value:
+                stripped_key = self._strip_prefix(inner_key, "tag", prefix_list)
+                if stripped_key in self.tag_keys:
+                    param_tag_keys.add(inner_key)
+            stripped_key = self._strip_prefix(key, "tag", prefix_list)
+            if stripped_key in self.tag_keys:
+                param_tag_keys.add(key)
+        self.tag_keys = param_tag_keys
```

Comment:
I don't quite understand why this was moved to here. Previously this chunk of code would not run if self.tag_keys was empty, now it runs regardless. I'd either switch this back, or add a check for 
```
if not self.tag_keys:
    return
```

Reply:
Yeah, the db model could contain no tag keys, so I will add a `if not self.tag_keys`

----
*TLDR:  Let me provide a little context to why I moved the logic here.*

So, I noticed we were doing the same if condition twice pretty much back to back inside of the init. Once inside of the [_set_tag_keys](https://github.com/project-koku/koku/blob/main/koku/api/query_params.py#L482) and once inside of [_validate](https://github.com/project-koku/koku/blob/main/koku/api/query_params.py#L482). 

I then noticed that the only time `_set_tag_keys()` is called is only called in the init, then we call the `_validate` function which also contains our only call of `_process_tag_query_params` as long as the same if statement is true. 

```
self._set_tag_keys()  # sets self.tag_keys
self._set_aws_category_keys()
self._validate(query_params)  # sets self.parameters
```

So, I then spent a little time trying to understand the purpose of the `_process_tag_query_params` function. When we called `_set_tag_keys` we use the if condition to see if we should collect all the acceptable tag keys from the database. I then realized  the purpose of the `_process_tag_query_params` was to reduce the tag key set to be only be keys passed into the `query_params`, but at the same time validate that those keys exist in the DB. Which is why we initially set the tag keys to all keys in the db in the `_set_tag_keys`. 

I thought this manipulation of the tag key set made more sense inside of the `_set_tag_keys` method. So, I removed `_process_tag_query_params` method and passed the query_params to the `_set_tag_keys`. I took the if condition short circuit returns for the `_set_tag_keys` function. 

I then changed the if condition to be a short circuit for the `_set_tag_keys`. If it tags or if we don't find a prefix in the url, exit and leave the tag keys as an empty set. Otherwise grab the tag keys from the db and check the query_param keys.